content/blog/insecure-software-development/index.md

```diff
@@ -0,0 +1,110 @@
+3. Code review is a great tool to reduce complexity before it becomes a problem
+4. Type juggling is still a thing
+
+## Introduction
+
+Do you remember `Log4J`? What about `Dependency Confusion`? How could you not? At the time of this writing both happened over this past year and disrupted parts of our industry. Development and Infrastructure teams scrambled to patch their systems. Companies with poor asset management had it even worse, trying to identify and uncover vulnerable infrastructure while responding to these incidents. So what does both of these incidents
+have in common? I will blame _hidden complexity_ here.
+
+Log4J abstracted environment variable parsing in log messages through a custom syntax that is parsed at the time of logging. Dependency confusion was caused by package managers trying to abstract the domain of dependency management, which is far from simple. In both cases library maintainers made the behaviour implicit.
+
+## Complexity
+
+As software evolves and gains more features, it becomes complex, with subtle dependencies across components. This leads to bugs and longer development cycles. Eventually no single developer can keep the entire context of their codebases in their heads anymore and more bugs start popping up, slowing the development cycle even more. There are usually two approaches to fight complexity:
+
+1. Make the code simpler and obvious
+2. Encapsulate complexity, abstract
+
+### Make the code simpler and obvious
+
+The idea here is to make things _explicit_. The less our users have to guess in order to use what we built the better. Users here can range from developers to end users of our products, libraries or our own codebases.
+
+1. Good naming practices, meaningful names
+   - Can we understand on a high level what this code is trying to achieve without having to dig deeper?
+2. Eliminating special cases
+   - Can we enforce invariants?
+   - Can we just crash instead of handling special cases?
+   - Can we have a single implementation for our interfaces?
+3. Ensuring consistency
+   - Are we following the same patterns across the codebase?
+   - Are we enforcing our conventions?
+   - Are our comments high level and useful?
+   - Do we have good defaults?
+
+From a security point of view fighting complexity by making the code simpler and obvious is a clear winner. Why should we encapsulate parsing environment variables from log messages when this is clearly a special case? Developers can explicitly handle this complexity instead of relying on the library.
+
+### Encapsulate complexity, abstract
+
+The idea of this approach is to _hide complexity_ so developers don't need to understand all the complexities of a certain domain in order to work on a codebase. This is usually related to classes or modules, where each one is designed to be relatively independent of each other. Developers can work in a module without concerning themselves with another. Abstractions can go wrong in two ways though:
```

Comment:
This goes against your previous point. In the previous point you're saying:
> The idea here is to make things _explicit_

which is contradictory to 

> The idea of this approach is to _hide complexity_

If I understand your intentions correctly, your previous point is mostly concerned with _implementation_ as opposed to this one which is mostly concerned with _interface_. i.e. I'm assuming you're saying we should provide enough information so that it's easy to understand and follow the code so that people know what they're doing, but also not concern people with unnecessary information to confuse them and have them shoot themselves in the foot and find the balance _correctly_.

I think adding a sentence to acknowledge this contradiction, pointing out the importance of correctly balancing which parts to make explicit and which parts to hide, and a mention of implementation vs interface could resolve this. 🤔

Reply:
Great point 👏